pkg/api/dashboard_public.go

```diff
@@ -54,43 +57,72 @@ func (hs *HTTPServer) GetPublicDashboardConfig(c *models.ReqContext) response.Re
 
 // sets public dashboard configuration for dashboard
 func (hs *HTTPServer) SavePublicDashboardConfig(c *models.ReqContext) response.Response {
-	pdc := &models.PublicDashboardConfig{}
-	if err := web.Bind(c.Req, pdc); err != nil {
+	pubdash := &models.PublicDashboard{}
+	if err := web.Bind(c.Req, pubdash); err != nil {
 		return response.Error(http.StatusBadRequest, "bad request data", err)
 	}
 
+	// Always set the org id to the current auth session orgId
+	pubdash.OrgId = c.OrgId
+
 	dto := dashboards.SavePublicDashboardConfigDTO{
-		OrgId:                 c.OrgId,
-		DashboardUid:          web.Params(c.Req)[":uid"],
-		PublicDashboardConfig: pdc,
+		OrgId:           c.OrgId,
+		DashboardUid:    web.Params(c.Req)[":uid"],
+		UserId:          c.UserId,
+		PublicDashboard: pubdash,
 	}
 
-	pdc, err := hs.dashboardService.SavePublicDashboardConfig(c.Req.Context(), &dto)
+	pubdash, err := hs.dashboardService.SavePublicDashboardConfig(c.Req.Context(), &dto)
 	if err != nil {
 		return handleDashboardErr(http.StatusInternalServerError, "Failed to save public dashboard configuration", err)
 	}
 
-	return response.JSON(http.StatusOK, pdc)
+	return response.JSON(http.StatusOK, pubdash)
 }
 
 // QueryPublicDashboard returns all results for a given panel on a public dashboard
-// POST /api/public/dashboard/:uid/panels/:panelId/query
+// POST /api/public/dashboard/:accessToken/panels/:panelId/query
 func (hs *HTTPServer) QueryPublicDashboard(c *models.ReqContext) response.Response {
 	panelId, err := strconv.ParseInt(web.Params(c.Req)[":panelId"], 10, 64)
 	if err != nil {
 		return response.Error(http.StatusBadRequest, "invalid panel ID", err)
 	}
 
+	dashboard, err := hs.dashboardService.GetPublicDashboard(c.Req.Context(), web.Params(c.Req)[":accessToken"])
+	if err != nil {
+		return response.Error(http.StatusInternalServerError, "could not fetch dashboard", err)
+	}
+
+	publicDashboard, err := hs.dashboardService.GetPublicDashboardConfig(c.Req.Context(), dashboard.OrgId, dashboard.Uid)
+	if err != nil {
+		return response.Error(http.StatusInternalServerError, "could not fetch public dashboard", err)
+	}
+
 	reqDTO, err := hs.dashboardService.BuildPublicDashboardMetricRequest(
 		c.Req.Context(),
-		web.Params(c.Req)[":uid"],
+		dashboard,
+		publicDashboard,
 		panelId,
 	)
 	if err != nil {
 		return handleDashboardErr(http.StatusInternalServerError, "Failed to get queries for public dashboard", err)
 	}
 
-	resp, err := hs.queryDataService.QueryDataMultipleSources(c.Req.Context(), nil, c.SkipCache, reqDTO, true)
+	// Get all needed datasource UIDs from queries
+	var uids []string
+	for _, query := range reqDTO.Queries {
+		uids = append(uids, query.Get("datasource").Get("uid").MustString())
+	}
+
+	// Create a temp user with read-only datasource permissions
+	anonymousUser := &models.SignedInUser{OrgId: dashboard.OrgId, Permissions: make(map[int64]map[string][]string)}
```

Comment:
We're creating an anonymous user with the dashboard OrgId, but really, this anonymous user should probably represent the user who created the public dashboard to keep intent. We could get this from PublicDashboard.CreatedBy. However, I think we need to check with auth team in the event that we start checking dashboard.OrgId == user.OrgId == datasource.OrgId

Reply:
Handle this later: https://github.com/grafana/grafana-enterprise-partnerships-team/issues/238